java/daikon/chicory/Instrument24.java

```diff
@@ -0,0 +1,1720 @@
+import java.lang.constant.ClassDesc;
+import java.lang.constant.ConstantDesc;
+import java.lang.constant.MethodTypeDesc;
+import java.lang.instrument.ClassFileTransformer;
+import java.lang.instrument.IllegalClassFormatException;
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.AccessFlag;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.ProtectionDomain;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.checker.signature.qual.BinaryName;
+import org.checkerframework.checker.signature.qual.ClassGetName;
+import org.checkerframework.checker.signature.qual.FieldDescriptor;
+import org.checkerframework.checker.signature.qual.FqBinaryName;
+import org.checkerframework.checker.signature.qual.InternalForm;
+import org.checkerframework.dataflow.qual.Pure;
+
+/**
+ * The Instrument24 class is responsible for modifying another class's bytecodes. Specifically, its
+ * main task is to add calls into the Chicory Runtime at method entries and exits for
+ * instrumentation purposes. These added calls are sometimes referred to as "hooks".
+ *
+ * <p>Starting with JDK 24, Java has added a set of APIs for reading and modifying .class files
+ * ({@code java.lang.classfile}).
+ *
+ * <p>We are migrating from BCEL to this new set of APIs for two main reasons:
+ *
+ * <ol>
+ *   <li>The new APIs are more complete and robust - no more fiddling with StackMaps.
+ *   <li>Since the new APIs are part of the official JDK release, they will always be up to date
+ *       with any .class file changes.
+ * </ol>
+ *
+ * <p>The files Instrument24.java and MethodGen24.java were added to Chicory to use this new set of
+ * APIs instead of BCEL. (We will need to continue to support Instrument.java using BCEL, as we do
+ * not anticipate our clients moving from JDK 8, 11, 17 or 21 to JDK 24 for quite some time.)
+ */
+public class Instrument24 implements ClassFileTransformer {
+
+  /** A log to which to print debugging information about program instrumentation. */
+  protected SimpleLog debugInstrument = new SimpleLog(false);
+
+  /** Directory for debug output. */
+  File debug_dir;
+
+  /** Directory into which to dump debug-instrumented classes. */
+  File debug_bin_dir;
+
+  /** Directory into which to dump original classes. */
+  File debug_orig_dir;
+
+  /** The index of this method in SharedData.methods. */
+  int cur_method_info_index = 0;
+
+  /** The MethodInfo for the current method. */
+  @Nullable MethodInfo curMethodInfo;
```

Comment:
I don't think this should be a field.  The value can be passed from `instrument_all_methods` to `instrumentCode`.